{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "94db8d0f-2dba-41af-a860-fcbe60c1824b",
   "metadata": {},
   "source": [
    "# Generate RTX-KG2 Metanames Parquet for Kuzu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e598413e-509b-48b9-aab3-1c703b364c09",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import json\n",
    "import pathlib\n",
    "import shutil\n",
    "from typing import Any, Dict, Generator, List, Literal\n",
    "\n",
    "import awkward as ak\n",
    "import duckdb\n",
    "import ijson\n",
    "import kuzu\n",
    "import pyarrow as pa\n",
    "import requests\n",
    "from genson import SchemaBuilder\n",
    "from pyarrow import parquet\n",
    "from rtx_kg2_functions import (\n",
    "    count_items_under_top_level_name,\n",
    "    drop_table_if_exists,\n",
    "    find_top_level_names,\n",
    "    generate_cypher_table_create_stmt_from_parquet_file,\n",
    "    parse_items_by_topmost_item_name,\n",
    "    parse_metadata_by_object_name,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "46ab8fed-b335-4d55-a73e-a6904bc0bc46",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Kuzu dir: data/kg2c_lite_2.8.4.full.with-metanames.dataset.kuzu\n"
     ]
    }
   ],
   "source": [
    "# set data to be used throughout notebook\n",
    "chunk_size = 500000\n",
    "data_dir = \"data\"\n",
    "parquet_dir = f\"{data_dir}/\"\n",
    "source_data_url = \"https://github.com/ncats/translator-lfs-artifacts/raw/main/files/kg2c_lite_2.8.4.json.gz\"\n",
    "target_extracted_sample_data = (\n",
    "    f\"{data_dir}/{pathlib.Path(source_data_url).name.replace('.json.gz', '.json')}\"\n",
    ")\n",
    "parquet_dir = target_extracted_sample_data.replace(\".json\", \".full.dataset.parquet\")\n",
    "parquet_metanames_dir = target_extracted_sample_data.replace(\n",
    "    \".json\", \".full.with-metanames.dataset.parquet\"\n",
    ")\n",
    "kuzu_dir = target_extracted_sample_data.replace(\n",
    "    \".json\", \".full.with-metanames.dataset.kuzu\"\n",
    ")\n",
    "target_extracted_sample_data_schema_file = target_extracted_sample_data.replace(\n",
    "    \".json\", \".schema.json\"\n",
    ")\n",
    "print(f\"Kuzu dir: {kuzu_dir}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3dc7f93e-e2ea-4840-868b-885afaff38f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create path for the kuzu database to reside\n",
    "if pathlib.Path(kuzu_dir).is_dir():\n",
    "    shutil.rmtree(kuzu_dir)\n",
    "pathlib.Path(kuzu_dir).mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "06079952-da05-431b-9c35-02487383943b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create path for parquet metanames to reside\n",
    "if pathlib.Path(parquet_metanames_dir).is_dir():\n",
    "    shutil.rmtree(parquet_metanames_dir)\n",
    "pathlib.Path(parquet_metanames_dir).mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b65ae82e-cdb2-4c10-a5a5-4114c327a06c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# init a Kuzu database and connection\n",
    "db = kuzu.Database(f\"{kuzu_dir}\")\n",
    "kz_conn = kuzu.Connection(db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "760b5a53-a85f-4334-ad39-dcd4807edabf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gather_table_names_from_parquet_path(\n",
    "    parquet_path: str,\n",
    "    column_with_table_name: str = \"id\",\n",
    "):\n",
    "    # return distinct table types as set comprehension\n",
    "    return set(\n",
    "        # create a parquet dataset and read a single column as an array\n",
    "        parquet.ParquetDataset(parquet_path)\n",
    "        .read(columns=[column_with_table_name])[column_with_table_name]\n",
    "        .to_pylist()\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "79d57098-70d4-4db8-bedf-8b087bfd7784",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exporting metaname tables for: biolink:Transcript\n",
      "Exporting metaname tables for: biolink:MicroRNA\n",
      "Exporting metaname tables for: biolink:Behavior\n",
      "Exporting metaname tables for: biolink:Drug\n",
      "Exporting metaname tables for: biolink:MolecularActivity\n",
      "Exporting metaname tables for: biolink:GrossAnatomicalStructure\n",
      "Exporting metaname tables for: biolink:Exon\n",
      "Exporting metaname tables for: biolink:ChemicalEntity\n",
      "Exporting metaname tables for: biolink:Phenomenon\n",
      "Exporting metaname tables for: biolink:CellLine\n",
      "Exporting metaname tables for: biolink:GenomicEntity\n",
      "Exporting metaname tables for: biolink:Device\n",
      "Exporting metaname tables for: biolink:Cell\n",
      "Exporting metaname tables for: biolink:LifeStage\n",
      "Exporting metaname tables for: biolink:Publication\n",
      "Exporting metaname tables for: biolink:EnvironmentalProcess\n",
      "Exporting metaname tables for: biolink:CellularComponent\n",
      "Exporting metaname tables for: biolink:Gene\n",
      "Exporting metaname tables for: biolink:NoncodingRNAProduct\n",
      "Exporting metaname tables for: biolink:OrganismTaxon\n",
      "Exporting metaname tables for: biolink:RetrievalSource\n",
      "Exporting metaname tables for: biolink:IndividualOrganism\n",
      "Exporting metaname tables for: biolink:PhenotypicFeature\n",
      "Exporting metaname tables for: biolink:BiologicalProcess\n",
      "Exporting metaname tables for: biolink:Procedure\n",
      "Exporting metaname tables for: biolink:GeneticInheritance\n",
      "Exporting metaname tables for: biolink:AnatomicalEntity\n",
      "Exporting metaname tables for: biolink:Disease\n",
      "Exporting metaname tables for: biolink:OrganismalEntity\n",
      "Exporting metaname tables for: biolink:OrganismAttribute\n",
      "Exporting metaname tables for: biolink:NamedThing\n",
      "Exporting metaname tables for: biolink:PopulationOfIndividualOrganisms\n",
      "Exporting metaname tables for: biolink:MaterialSample\n",
      "Exporting metaname tables for: biolink:GeographicLocation\n",
      "Exporting metaname tables for: biolink:ProteinFamily\n",
      "Exporting metaname tables for: biolink:Food\n",
      "Exporting metaname tables for: biolink:PathologicalProcess\n",
      "Exporting metaname tables for: biolink:Protein\n",
      "Exporting metaname tables for: biolink:Polypeptide\n",
      "Exporting metaname tables for: biolink:PhysicalEntity\n",
      "Exporting metaname tables for: biolink:SmallMolecule\n",
      "Exporting metaname tables for: biolink:Cohort\n",
      "Exporting metaname tables for: biolink:GeneFamily\n",
      "Exporting metaname tables for: biolink:Treatment\n",
      "Exporting metaname tables for: biolink:Event\n",
      "Exporting metaname tables for: biolink:BehavioralFeature\n",
      "Exporting metaname tables for: biolink:BiologicalEntity\n",
      "Exporting metaname tables for: biolink:EnvironmentalFeature\n",
      "Exporting metaname tables for: biolink:ProteinDomain\n",
      "Exporting metaname tables for: biolink:Agent\n",
      "Exporting metaname tables for: biolink:MolecularMixture\n",
      "Exporting metaname tables for: biolink:ClinicalAttribute\n",
      "Exporting metaname tables for: biolink:Activity\n",
      "Exporting metaname tables for: biolink:PhysiologicalProcess\n",
      "Exporting metaname tables for: biolink:NucleicAcidEntity\n",
      "Exporting metaname tables for: biolink:ComplexMolecularMixture\n",
      "Exporting metaname tables for: biolink:ChemicalMixture\n",
      "Exporting metaname tables for: biolink:RNAProduct\n",
      "Exporting metaname tables for: biolink:DiseaseOrPhenotypicFeature\n",
      "Exporting metaname tables for: biolink:Pathway\n",
      "Exporting metaname tables for: biolink:ClinicalIntervention\n",
      "Exporting metaname tables for: biolink:MolecularEntity\n",
      "Exporting metaname tables for: biolink:InformationContentEntity\n",
      "Exporting metaname tables for: biolink:has_completed\n",
      "Exporting metaname tables for: biolink:manifestation_of\n",
      "Exporting metaname tables for: biolink:physically_interacts_with\n",
      "Exporting metaname tables for: biolink:capable_of\n",
      "Exporting metaname tables for: biolink:has_output\n",
      "Exporting metaname tables for: biolink:produces\n",
      "Exporting metaname tables for: biolink:has_not_completed\n",
      "Exporting metaname tables for: biolink:derives_from\n",
      "Exporting metaname tables for: biolink:has_decreased_amount\n",
      "Exporting metaname tables for: biolink:close_match\n",
      "Exporting metaname tables for: biolink:has_increased_amount\n",
      "Exporting metaname tables for: biolink:precedes\n",
      "Exporting metaname tables for: biolink:related_to\n",
      "Exporting metaname tables for: biolink:overlaps\n",
      "Exporting metaname tables for: biolink:has_plasma_membrane_part\n",
      "Exporting metaname tables for: biolink:occurs_in\n",
      "Exporting metaname tables for: biolink:treats\n",
      "Exporting metaname tables for: biolink:predisposes\n",
      "Exporting metaname tables for: biolink:transcribed_from\n",
      "Exporting metaname tables for: biolink:coexists_with\n",
      "Exporting metaname tables for: biolink:actively_involved_in\n",
      "Exporting metaname tables for: biolink:located_in\n",
      "Exporting metaname tables for: biolink:has_participant\n",
      "Exporting metaname tables for: biolink:has_metabolite\n",
      "Exporting metaname tables for: biolink:has_molecular_consequence\n",
      "Exporting metaname tables for: biolink:colocalizes_with\n",
      "Exporting metaname tables for: biolink:chemically_similar_to\n",
      "Exporting metaname tables for: biolink:biomarker_for\n",
      "Exporting metaname tables for: biolink:enables\n",
      "Exporting metaname tables for: biolink:homologous_to\n",
      "Exporting metaname tables for: biolink:has_phenotype\n",
      "Exporting metaname tables for: biolink:prevents\n",
      "Exporting metaname tables for: biolink:exacerbates\n",
      "Exporting metaname tables for: biolink:expressed_in\n",
      "Exporting metaname tables for: biolink:associated_with\n",
      "Exporting metaname tables for: biolink:gene_associated_with_condition\n",
      "Exporting metaname tables for: biolink:same_as\n",
      "Exporting metaname tables for: biolink:model_of\n",
      "Exporting metaname tables for: biolink:diagnoses\n",
      "Exporting metaname tables for: biolink:develops_from\n",
      "Exporting metaname tables for: biolink:subclass_of\n",
      "Exporting metaname tables for: biolink:disease_has_location\n",
      "Exporting metaname tables for: biolink:disrupts\n",
      "Exporting metaname tables for: biolink:gene_product_of\n",
      "Exporting metaname tables for: biolink:causes\n",
      "Exporting metaname tables for: biolink:affects\n",
      "Exporting metaname tables for: biolink:regulates\n",
      "Exporting metaname tables for: biolink:has_part\n",
      "Exporting metaname tables for: biolink:disease_has_basis_in\n",
      "Exporting metaname tables for: biolink:composed_primarily_of\n",
      "Exporting metaname tables for: biolink:temporally_related_to\n",
      "Exporting metaname tables for: biolink:contributes_to\n",
      "Exporting metaname tables for: biolink:is_sequence_variant_of\n",
      "Exporting metaname tables for: biolink:translates_to\n",
      "Exporting metaname tables for: biolink:indirectly_physically_interacts_with\n",
      "Exporting metaname tables for: biolink:catalyzes\n",
      "Exporting metaname tables for: biolink:has_input\n",
      "Exporting metaname tables for: biolink:directly_physically_interacts_with\n",
      "Exporting metaname tables for: biolink:lacks_part\n",
      "Exporting metaname tables for: biolink:mentions\n",
      "Exporting metaname tables for: biolink:contraindicated_for\n",
      "Exporting metaname tables for: biolink:in_taxon\n",
      "Exporting metaname tables for: biolink:has_member\n",
      "Exporting metaname tables for: biolink:superclass_of\n",
      "Exporting metaname tables for: biolink:correlated_with\n"
     ]
    }
   ],
   "source": [
    "for path, table_name_column in [\n",
    "    [f\"{parquet_dir}/nodes\", \"category\"],\n",
    "    [f\"{parquet_dir}/edges\", \"predicate\"],\n",
    "]:\n",
    "\n",
    "    # create object base dir\n",
    "    parquet_metanames_object_base = f\"{parquet_metanames_dir}/{pathlib.Path(path).name}\"\n",
    "    pathlib.Path(parquet_metanames_object_base).mkdir(exist_ok=True)\n",
    "\n",
    "    for table_name in gather_table_names_from_parquet_path(\n",
    "        parquet_path=path, column_with_table_name=table_name_column\n",
    "    ):\n",
    "        # create metanames / objects using cypher safe name and dir\n",
    "        cypher_safe_table_name = table_name.split(\":\")[1]\n",
    "        parquet_metanames_metaname_base = (\n",
    "            f\"{parquet_metanames_object_base}/{cypher_safe_table_name}\"\n",
    "        )\n",
    "        pathlib.Path(parquet_metanames_metaname_base).mkdir(exist_ok=True)\n",
    "\n",
    "        print(f\"Exporting metaname tables for: {table_name}\")\n",
    "\n",
    "        # determine rowcount for offsetting parquet files by metaname\n",
    "        with duckdb.connect() as ddb:\n",
    "            rowcount = int(\n",
    "                ddb.execute(\n",
    "                    f\"\"\"\n",
    "                    SELECT COUNT(*) as count\n",
    "                    FROM read_parquet('{path}/*')\n",
    "                    WHERE {table_name_column}='{table_name}';\n",
    "                    \"\"\"\n",
    "                ).fetchone()[0]\n",
    "            )\n",
    "\n",
    "        # create a chunk offsets list for chunked parquet output by metaname\n",
    "        chunk_offsets = list(\n",
    "            range(\n",
    "                0,\n",
    "                # gather rowcount from table and use as maximum for range\n",
    "                rowcount,\n",
    "                # step through using chunk size\n",
    "                chunk_size,\n",
    "            )\n",
    "        )\n",
    "\n",
    "        # gather chunks of data by metaname and export to file using chunked offsets\n",
    "        with duckdb.connect() as ddb:\n",
    "            for idx, offset in enumerate(chunk_offsets):\n",
    "                ddb.execute(\n",
    "                    f\"\"\"\n",
    "                    COPY (\n",
    "                        SELECT *\n",
    "                        FROM read_parquet('{path}/*')\n",
    "                        WHERE {table_name_column}='{table_name}'\n",
    "                        LIMIT {chunk_size} OFFSET {offset}\n",
    "                    )\n",
    "                    TO '{parquet_metanames_metaname_base}/{cypher_safe_table_name}.{idx}.parquet' (FORMAT PARQUET);\n",
    "                    \"\"\"\n",
    "                )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4c1351e9-5c8b-43a6-9ead-8f554f1fcb26",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Equal values from parquet source data/kg2c_lite_2.8.4.full.dataset.parquet/nodes/* to metanames target data/kg2c_lite_2.8.4.full.with-metanames.dataset.parquet/nodes/**\n",
      "Equal values from parquet source data/kg2c_lite_2.8.4.full.dataset.parquet/edges/* to metanames target data/kg2c_lite_2.8.4.full.with-metanames.dataset.parquet/edges/**\n"
     ]
    }
   ],
   "source": [
    "def gather_parquet_dataset_size_from_path(parquet_path: str):\n",
    "    with duckdb.connect() as ddb:\n",
    "        return int(\n",
    "            ddb.execute(\n",
    "                f\"\"\"\n",
    "                SELECT COUNT(*) as count\n",
    "                FROM read_parquet('{parquet_path}')\n",
    "                \"\"\"\n",
    "            ).fetchone()[0]\n",
    "        )\n",
    "\n",
    "\n",
    "for source, target in [\n",
    "    [\n",
    "        \"data/kg2c_lite_2.8.4.full.dataset.parquet/nodes/*\",\n",
    "        \"data/kg2c_lite_2.8.4.full.with-metanames.dataset.parquet/nodes/**\",\n",
    "    ],\n",
    "    [\n",
    "        \"data/kg2c_lite_2.8.4.full.dataset.parquet/edges/*\",\n",
    "        \"data/kg2c_lite_2.8.4.full.with-metanames.dataset.parquet/edges/**\",\n",
    "    ],\n",
    "]:\n",
    "    if gather_parquet_dataset_size_from_path(\n",
    "        source\n",
    "    ) != gather_parquet_dataset_size_from_path(target):\n",
    "        raise ValueError(\n",
    "            f\"Inequal number of rows from parquet source {source} to metanames target {target}.\"\n",
    "        )\n",
    "    else:\n",
    "        print(f\"Equal values from parquet source {source} to metanames target {target}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
